pkg/storage/s3.go

```diff
@@ -1,56 +1,318 @@
+	options *backup.S3
+}
+
+// S3BackendOptions contains options for s3 storage
 type S3BackendOptions struct {
-	Endpoint string `json:"endpoint" toml:"endpoint"`
-	Region   string `json:"region" toml:"region"`
+	Endpoint              string `json:"endpoint" toml:"endpoint"`
+	Region                string `json:"region" toml:"region"`
+	StorageClass          string `json:"storage_class" toml:"storage_class"`
+	SSE                   string `json:"sse" toml:"sse"`
+	ACL                   string `json:"acl" toml:"acl"`
+	AccessKey             string `json:"access_key" toml:"access_key"`
+	SecretAccessKey       string `json:"secret_access_key" toml:"secret_access_key"`
+	Provider              string `json:"provider" toml:"provider"`
+	ForcePathStyle        bool   `json:"force_path_style" toml:"force_path_style"`
+	UseAccelerateEndpoint bool   `json:"use_accelerate_endpoint" toml:"use_accelerate_endpoint"`
 }
 
 func (options *S3BackendOptions) apply(s3 *backup.S3) error {
 	if options.Endpoint == "" && options.Region == "" {
 		return errors.New("must provide either 's3.region' or 's3.endpoint'")
 	}
-	// TODO: Verify Region.
+	if options.Endpoint != "" {
+		if !strings.HasPrefix(options.Endpoint, "https://") &&
+			!strings.HasPrefix(options.Endpoint, "http://") {
+			options.Endpoint = "http://" + options.Endpoint
+		}
+	}
+	if options.Provider == "alibaba" || options.Provider == "netease" ||
+		options.UseAccelerateEndpoint {
+		options.ForcePathStyle = false
+	}
+	if options.AccessKey == "" && options.SecretAccessKey != "" {
+		return errors.New("access_key not found")
+	}
+	if options.AccessKey != "" && options.SecretAccessKey == "" {
+		return errors.New("secret_access_key not found")
+	}
 
 	s3.Endpoint = options.Endpoint
 	s3.Region = options.Region
+	// StorageClass, SSE and ACL are acceptable to be empty
+	s3.StorageClass = options.StorageClass
+	s3.Sse = options.SSE
+	s3.Acl = options.ACL
+	s3.AccessKey = options.AccessKey
+	s3.SecretAccessKey = options.SecretAccessKey
+	s3.ForcePathStyle = options.ForcePathStyle
 	return nil
 }
 
 func defineS3Flags(flags *pflag.FlagSet) {
-	flags.String(s3EndpointOption, "", "Set the AWS S3 endpoint URL")
-	flags.String(s3RegionOption, "", "Set the AWS region")
-	// TODO: Finalize the list of options.
-	_ = flags.MarkHidden(s3EndpointOption)
+	flags.String(s3EndpointOption, "", "Set the S3 endpoint URL")
+	flags.String(s3RegionOption, "", "Set the S3 region")
+	flags.String(s3StorageClassOption, "", "Set the S3 storage class")
+	flags.String(s3SSEOption, "", "Set the S3 server-side encryption algorithm")
+	flags.String(s3ACLOption, "", "Set the S3 canned ACLs")
+	flags.String(s3ProviderOption, "", "Set the S3 provider")
 }
 
 func getBackendOptionsFromS3Flags(flags *pflag.FlagSet) (options S3BackendOptions, err error) {
+	send, err := flags.GetBool(flagSendCredentialOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	if send {
+		options.AccessKey = os.Getenv("AWS_ACCESS_KEY_ID")
```

Comment:
Is there any way to get these credentials from the chained credentials? Because this is confusing, if its behavior not matching the AWS sdk.

Reply:
Updated.

Reply:
We could [call `.Get()`](https://godoc.org/github.com/aws/aws-sdk-go/aws/credentials#Credentials.Get) on any credentials to obtain the [tokens](https://godoc.org/github.com/aws/aws-sdk-go/aws/credentials#Value).

The problem is that non-static credentials can expire, the tokens obtained at the beginning may be useless after a while, breaking the BR process.

I think we may want to extract the tokens fields _out_ of StorageBackend, and recompute the tokens for every gRPC call of Backup/Download, if we want to support sending non-static credentials.

```proto
message BackupRequest {
    ...
    StorageBackend storage_backend = 9;
    AuthSecrets auth_secrets = 10;   // <-- new
}

message S3 {
    ...
    reserved 8 to 9;  // remove keys
}

message AuthSecrets {
    oneof secrets {
        S3AuthSecrets s3 = 3;
        GCSAuthSecrets gcs = 4;
    }
}

message S3AuthSecrets {
    string access_key = 8;
    string secret_access_key = 9;
    string session_token = 10;
}
```